src/api/oauth2.rs

```diff
@@ -0,0 +1,293 @@
+    };
+    let socket_addr = match url.socket_addrs(|| None) {
+        Ok(mut addrs) => addrs.pop(),
+        _ => None,
+    };
+    if let Some(s) = socket_addr {
+        if s.ip().is_loopback() {
+            return socket_addr;
+        }
+    }
+    None
+}
+
+/// Obtain a Spotify access token using the authorization code with PKCE OAuth flow.
+/// The redirect_uri must match what is registered to the client ID.
+pub fn get_access_token(
+    client_id: &str,
+    redirect_uri: &str,
+    scopes: Vec<&str>,
+) -> Result<OAuthToken, OAuthError> {
+    let auth_url = AuthUrl::new("https://accounts.spotify.com/authorize".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let token_url = TokenUrl::new("https://accounts.spotify.com/api/token".to_string())
+        .map_err(|_| OAuthError::InvalidSpotifyUri)?;
+    let redirect_url =
+        RedirectUrl::new(redirect_uri.to_string()).map_err(|e| OAuthError::InvalidRedirectUri {
+            uri: redirect_uri.to_string(),
+            e,
+        })?;
+    let client = BasicClient::new(
+        ClientId::new(client_id.to_string()),
+        None,
+        auth_url,
+        Some(token_url),
+    )
+    .set_redirect_uri(redirect_url);
+
+    let (pkce_challenge, pkce_verifier) = PkceCodeChallenge::new_random_sha256();
+
+    // Generate the full authorization URL.
+    // Some of these scopes are unavailable for custom client IDs. Which?
+    let request_scopes: Vec<oauth2::Scope> = scopes
+        .clone()
+        .into_iter()
+        .map(|s| Scope::new(s.into()))
+        .collect();
+    let (auth_url, _) = client
+        .authorize_url(CsrfToken::new_random)
+        .add_scopes(request_scopes)
+        .set_pkce_challenge(pkce_challenge)
+        .url();
+
+    println!("Browse to: {}", auth_url);
+    if let Err(err) = open::that(auth_url.to_string()) {
+        eprintln!("An error occurred when opening '{}': {}", auth_url, err)
+    }
+
+    let code = match get_socket_address(redirect_uri) {
+        Some(addr) => get_authcode_listener(addr),
+        _ => get_authcode_stdin(),
+    }?;
+    trace!("Exchange {code:?} for access token");
+
+    // Do this sync in another thread because I am too stupid to make the async version work.
+    let (tx, rx) = mpsc::channel();
+    std::thread::spawn(move || {
+        let resp = client
+            .exchange_code(code)
+            .set_pkce_verifier(pkce_verifier)
+            .request(http_client);
+        if let Err(e) = tx.send(resp) {
+            error!("OAuth channel send error: {e}");
+        }
+    });
+    let token_response = rx.recv().map_err(|_| OAuthError::Recv)?;
+    let token = token_response.map_err(|e| OAuthError::ExchangeCode { e: e.to_string() })?;
+    trace!("Obtained new access token: {token:?}");
+
+    let token_scopes: Vec<String> = match token.scopes() {
+        Some(s) => s.iter().map(|s| s.to_string()).collect(),
+        _ => scopes.into_iter().map(|s| s.to_string()).collect(),
```

Comment:
i guess its technically nullable but not really in practice, maybe it should be an error as well?

Reply:
fixed